### ISSUE TITLE
bug: `packageJson.exports.types` should be always be ordered first

### ISSUE THREAD

**Issue by jlarmstrongiv** [keithamus/sort-package-json#294]
Failing Example:

```json
{
  "name": "rcva",
  "version": "1.0.6",
  "private": false,
  "description": "React Class Variance Authority 🧬",
  "keywords": [
    "React Class Variance Authority",
    "react-class-variance-authority",
    "classes",
    "classname",
    "classnames",
    "css",
    "rcva",
    "stitches",
    "vanilla-extract",
    "variants",
    "windstitch"
  ],
  "homepage": "https://github.com/jlarmstrongiv/rcva/blob/main/packages/core/README.md",
  "bugs": "https://github.com/jlarmstrongiv/rcva/issues",
  "repository": "https://github.com/jlarmstrongiv/rcva.git",
  "license": "MIT",
  "author": "John L. Armstrong IV",
  "sideEffects": false,
  "type": "module",
  "exports": {
    ".": {
      "require": "./dist/index.cjs",
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    },
    "./package.json": "./package.json"
  },
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "source": "src/index.tsx",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup-node"
  },
  "devDependencies": {
    "@types/react": "^18.0.31",
    "@types/react-dom": "^18.0.11",
    "rimraf": "^4.4.1",
    "tsup": "^6.7.0",
    "typescript": "^5.0.0"
  },
  "peerDependencies": {
    "@leafygreen-ui/polymorphic": "^1.3.0",
    "class-variance-authority": ">= 0.5.1 < 1",
    "react": "^18.0.0",
    "react-dom": "^18.0.0",
    "tailwind-merge": "^1.11.0"
  }
}
```

More specifically, `Should be the first in the object as required by TypeScript.` according to [publint.dev](https://publint.dev/):

```json
{
  "exports": {
    ".": {
      "require": "./dist/index.cjs",
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    },
    "./package.json": "./package.json"
  }
}
```

**`EXPORTS_TYPES_SHOULD_BE_FIRST`**

Ensure types condition to be the first. The [TypeScript docs](https://www.typescriptlang.org/docs/handbook/esm-node.html#packagejson-exports-imports-and-self-referencing) recommends so, but it's also because the exports field is order-based.

For example, a scenario where both the types and import condition could be active, types should be first so that it matches and returns a .d.ts file, rather than a .js file from the import condition.

**Comment by keithamus** [comment]
PRs welcome!

**Comment by jlarmstrongiv** [comment]
@keithamus would the PR include these changes?
- adding documentation to [defaultRules](https://github.com/keithamus/sort-package-json/blob/main/defaultRules.md)
- editing the [`exports`](https://github.com/keithamus/sort-package-json/blob/main/index.js#L214) key to something like `{ key: 'exports', over: sortObjectBy(['types']) }`
- patch version bump

Let me know if I missed anything in the outline 😄 if not, it should be a straightforward fix

**Comment by keithamus** [comment]
That’s correct. We use semantic versioning so prefixing your commit `fix: ` will automatically apply and release the patch version.

**Comment by fisker** [comment]
If we are doing this, should we also enforce the `default` to be the last one?

**Comment by mrmckeb** [comment]
@keithamus are you still working on this? If not, I might pick it up.

**Comment by keithamus** [comment]
I am happy to accept PRs. I no longer contribute code to this repo as I prefer to let the community do so.

**Comment by mrmckeb** [comment]
Apologies, I meant to tag @jlarmstrongiv! I assume that this is not being worked on though?

**Comment by GeorgeTaveras1231** [comment]
@mrmckeb @jlarmstrongiv  I opened a PR (#336) addressing this - Just thought I'd let you know, I didn't see your comments about potentially working on this before I opened the PR.

In my first comment I suggested not sorting unless specified - my PR takes a more opinionated position on sorting known conditions while leaving unknown conditions in place (mostly).

**Comment by mrmckeb** [comment]
Absolutely fine, this had fallen into my backlog - thank you, @GeorgeTaveras1231!

**Comment by cefn** [comment]
I don't agree this feature is appropriate even for an opinionated formatter. 

It just broke all packages in our monorepo via `prettier-plugin-packagejson`. Like all monorepos following the recommendations of [this article](https://colinhacks.com/essays/live-types-typescript-monorepo) we have a custom condition that takes precedence. 

I consider this to be a pretty major regression meaning we have to back out our prettier tooling. Given this and related threads seem to be `argumentum ad verecundiam` note the recommendation comes from the author of [zod](https://www.npmjs.com/package/zod), who knows something about typescript. The approach is absolutely core to managing large and complex typescript codebases.

I quote...

> You should put your custom condition first. Order matters! It's important that your custom condition is the first one in the list (even before "types").

> Overall, my recommendation is `#5:` custom export conditions.
> It's clean, easy to configure, and works well with modern tooling.
> It doesn't require you to use pnpm, which is a non-starter for many projects.
> You don't need to worry about keeping runtime and TypeScript configurations in sync: you just set "exports" once using your custom condition, then tell your other tools (including TypeScript itself) to pay attention to that condition.

See also https://github.com/keithamus/sort-package-json/issues/294#issuecomment-2685900403 above which anticipated this breaking change but rather than leaving exports alone, I have a feeling this somehow prompted MORE opinions about ordering instead. 

I can't imagine how this issue can be legislated for unless you know every project in the world, meaning the only option is leaving well alone when order-is-configuration.
